src/bun.js/bindings/Script.cpp

```diff
@@ -0,0 +1,181 @@
+}
+
+JSC_DEFINE_CUSTOM_GETTER(scriptGetCachedDataRejected, (JSGlobalObject* globalObject, EncodedJSValue thisValue, PropertyName))
+{
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    return throwVMTypeError(globalObject, scope, "TODO"_s);
+}
+JSC_DEFINE_HOST_FUNCTION(scriptCreateCachedData, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    return throwVMTypeError(globalObject, scope, "TODO"_s);
+}
+JSC_DEFINE_HOST_FUNCTION(scriptRunInContext, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    return throwVMTypeError(globalObject, scope, "TODO"_s);
+}
+JSC_DEFINE_CUSTOM_GETTER(scriptGetSourceMapURL, (JSGlobalObject* globalObject, EncodedJSValue thisValue, PropertyName))
+{
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    return throwVMTypeError(globalObject, scope, "TODO"_s);
+}
+
+class ScriptPrototype final : public JSNonFinalObject {
+public:
+    using Base = JSNonFinalObject;
+
+    static ScriptPrototype* create(VM& vm, JSGlobalObject* globalObject, Structure* structure)
+    {
+        ScriptPrototype* ptr = new (NotNull, allocateCell<ScriptPrototype>(vm)) ScriptPrototype(vm, globalObject, structure);
+        ptr->finishCreation(vm, globalObject);
+        return ptr;
+    }
+
+    DECLARE_INFO;
+    template<typename CellType, SubspaceAccess>
+    static GCClient::IsoSubspace* subspaceFor(VM& vm)
+    {
+        return &vm.plainObjectSpace();
+    }
+    static Structure* createStructure(VM& vm, JSGlobalObject* globalObject, JSValue prototype)
+    {
+        return Structure::create(vm, globalObject, prototype, TypeInfo(ObjectType, StructureFlags), info());
+    }
+
+private:
+    ScriptPrototype(VM& vm, JSGlobalObject* globalObject, Structure* structure)
+        : Base(vm, structure)
+    {
+    }
+
+    void finishCreation(VM&, JSGlobalObject*);
+};
+STATIC_ASSERT_ISO_SUBSPACE_SHARABLE(ScriptPrototype, ScriptPrototype::Base);
+
+static const struct HashTableValue scriptPrototypeTableValues[] = {
+   { "cachedDataRejected"_s, static_cast<unsigned>(PropertyAttribute::ReadOnly|PropertyAttribute::CustomAccessor), NoIntrinsic, { HashTableValue::GetterSetterType, scriptGetCachedDataRejected, nullptr } },
+   { "createCachedData"_s, static_cast<unsigned>(PropertyAttribute::ReadOnly|PropertyAttribute::Function), NoIntrinsic, { HashTableValue::NativeFunctionType, scriptCreateCachedData, 0 } },
+   { "runInContext"_s, static_cast<unsigned>(PropertyAttribute::ReadOnly|PropertyAttribute::Function), NoIntrinsic, { HashTableValue::NativeFunctionType, scriptRunInContext, 0 } },
+   { "sourceMapURL"_s, static_cast<unsigned>(PropertyAttribute::ReadOnly|PropertyAttribute::CustomAccessor), NoIntrinsic, { HashTableValue::GetterSetterType, scriptGetSourceMapURL, nullptr } },
```

Comment:
btw, to implement this, have a look at `JSC::FunctionExecutable`. I vaguely remember something in there to extract the `sourceMapURL`

Reply:
One question I had here was whether I should be using a custom property here or just set the field on the object during init.
I feel like the latter might be simpler, but I'm not sure exactly how to do it - if you have any examples of code that sets object fields you could link that'd be great :)

Reply:
imo we should keep it as a custom getter/setter (present state) 

as an optional optimization, if we think people will frequently call it, then we could add a `mutable WriteBarrier<JSC::JSString> cachedSourceMapURL;` on the `Script` to cache it 

The `sourceMapURL` may be lazily loaded in JSC (i haven't checked), in which case making it a property on the object would be less efficient